domain-models-runtime/src/main/resources/templates/db_scripts/audit.ftl

```diff
@@ -10,29 +10,28 @@ CREATE OR REPLACE FUNCTION ${myuniversity}_${mymodule}.audit_${table.tableName}_
   <#if (table.auditingSnippet.insert.declare)??>
     ${table.auditingSnippet.insert.declare}
   </#if>
-    seed TEXT;
-    maxid UUID;
     jsonb JSONB;
+    uuidtext TEXT;
+    uuid UUID;
   BEGIN
-    maxid = (SELECT ${myuniversity}_${mymodule}.max(id) FROM ${myuniversity}_${mymodule}.${table.auditingTableName});
-    IF maxid IS NULL THEN
-      -- don't use random, that will be different on each pg-pool replication node
-      seed = md5(concat('${myuniversity}_${mymodule}.${table.auditingTableName}', NEW.jsonb));
-      -- avoid overflow
-      IF (left(seed, 13) = 'ffffffff-ffff') THEN
-        seed = overlay(seed placing '00000000-0000' from 1);
-      END IF;
-      -- UUID version byte
-      seed = overlay(seed placing '4' from 13);
-      -- UUID variant byte
-      seed = overlay(seed placing '8' from 17);
-      maxid = seed::uuid;
-    ELSE
-      maxid = ${myuniversity}_${mymodule}.next_uuid(maxid);
-    END IF;
+    jsonb = CASE WHEN TG_OP = 'DELETE' THEN OLD.jsonb ELSE NEW.jsonb END;
+
+    -- create uuid based on the jsonb value so that concurrent updates of different records are possible.
+    uuidtext = md5(jsonb::text);
+    -- UUID version byte
+    uuidtext = overlay(uuidtext placing '4' from 13);
+    -- UUID variant byte
+    uuidtext = overlay(uuidtext placing '8' from 17);
+    uuid = uuidtext::uuid;
+    -- If uuid is already in use increment until an unused is found. This can only happen if the jsonb content
+    -- is exactly the same. This should be very rare when it includes a timestamp.
+    WHILE EXISTS (SELECT 1 FROM ${myuniversity}_${mymodule}.${table.auditingTableName} WHERE id = uuid) LOOP
+      uuid = ${myuniversity}_${mymodule}.next_uuid(uuid);
```

Comment:
If concurrency happens here, will they get the same uuid as well (the new uuid will not be saved/committed yet)?

Reply:
This can only happen if the two transactions try to alter the same record. Then one of them will rollback even if there is no audit table.
RMB-430 is to fix the case where the two concurrent transactions alter different records.